### PR TITLE
@joystream/types tooling improvements - augment-codec and defs.json

### PR DESCRIPTION
**Pull request by Lezek123**
Draft based on https://github.com/Joystream/joystream/pull/1433
Implements: https://github.com/Joystream/joystream/issues/1393

Changes:
- Full augmentation support for custom types (`@joystream/types/augment-codec`) - replaced `generateCodecDefs` script with `generateAugmentCodec`
- Modified `generateRegistryJson` script to create perfectly valid `defs.json`. Tweaks to make it compatible with `@polkadot/typegen` are implemented at later stage (in `types/augment/all/definitions.ts`), which should make it possible to use `defs.json` file to register joystream types without any issues (this may be helpful for applications that expect types in json string format or for applications that want to use `@joystream/types/augment` with 100% compatibility)

Unfortunately it turned out that using the new augmentation in Pioneer will still be very tricky, because packages like `@polkadot/react-signer` etc. are often trying to access modules/methods that don't exist in Joystream chain (ie. `api.tx.proxy.proxy`), resulting in a lot of typescript errors in those packages. To some extend this provides valueable information, but fixing all those errors would take a lot of time and probably make it very hard or almost impossible to pull from upstream `@polkadot-js/apps` again (if we ever plan to do this)

A potential solution may be to have an additional signature in augmented apis like:
```
    [module: string]: {
      [method: string]: QueryableStorageEntry<ApiType>;
    },
``` 
So that, for example, any unknown `api.query` method fallsback to `QueryableStorageEntry<ApiType>`.
Further investigation required.

### CODE REVIEW

File: types/src/mint/index.ts [hunk 1]
```diff
@@ -49,10 +51,10 @@ export class Mint
   implements IMint {}
 
 export const mintTypes: RegistryTypes = {
-  MintId: 'u64',
+  MintId,
   Mint,
-  MintBalanceOf: 'Balance',
-  BalanceOfMint: 'Balance',
```
Comment (mnaamani) [inline types/src/mint/index.ts:55]: Strictly speaking we want the underlying u128 to be inferred by the type of the Balance in the system. It is u128 but that is configured in the runtime. I doubt we will change it but as a general point how can we avoid explicitly using u128 here?

All currencies in our runtime use this type: https://github.com/Joystream/joystream/blob/master/runtime-modules/common/src/currency.rs#L11

The Balance trait type comes from the balances module:
https://github.com/Joystream/joystream/blob/master/runtime/src/lib.rs#L215

The primitive type for Balance is defined here: https://github.com/Joystream/joystream/blob/master/runtime/src/primitives.rs#L33

**Comment by mnaamani** [comment]
Which version of the runtime did you run the `yarn generate:all` against? should i do it against the runtime built from `content_directory_second_try` branch?

**Comment by Lezek123** [comment]
> Which version of the runtime did you run the `yarn generate:all` against? should i do it against the runtime built from `content_directory_second_try` branch?

Yes, I used `content_directory_second_try` branch.

**Comment by mnaamani** [comment]
Ok great I've run the augment generate scripts and I'm getting the same results.

Looks good.

> Unfortunately it turned out that using the new augmentation in Pioneer will still be very tricky, because packages like `@polkadot/react-signer` etc. are often trying to access modules/methods that don't exist in Joystream chain

Thats too bad. There are a couple of modules that we might consider adding like Proxy and Multisig to our runtime because we could end up using them, but unfortunately it doesn't make sense for many of the other modules which we haven't included in our runtime. 

I can't say I understand exactly how you would implement the suggested solution for pioneer, but clearly if the fallback works then it is just as good as when we didn't have any of these definitions, and we at least cover the apis and types we do have in our runtime. 

It is certainly not worth the effort to try to fix the  typescript errors. I would rather we are able to pull from upstream if/when we need to.

So with this change I take it, there is no need to update the paths in tsconfig.json files of dependant packages, since the generated files are still in the same location? `types/augment-codec`

And pardon my ignorance, but why is there also a similar `types/augment/` and when would that be used instead of `types/augment-codec/` ?

**Comment by Lezek123** [comment]
> Thats too bad. There are a couple of modules that we might consider adding like Proxy and Multisig to our runtime because we could end up using them, but unfortunately it doesn't make sense for many of the other modules which we haven't included in our runtime.

We could exclude some packages we don't need through `tsconfig` and `.eslintignore`, but this still leaves us with quite a few issues. I'll try to think about it, maybe I'll be able to find some quick and reasonable solution.

> So with this change I take it, there is no need to update the paths in tsconfig.json files of dependant packages, since the generated files are still in the same location? types/augment-codec

Yes, but now some projects may also want to add:
```
      "@polkadot/api/augment": ["path/to/types/augment-codec/augment-api.ts"]
```
to get the api qu(which I initially planned to do in Pioneer)

> And pardon my ignorance, but why is there also a similar types/augment/ and when would that be used instead of types/augment-codec/ ?

Inside `augment` we have the augmentation that `@polkadot/typegen` tool generates. It takes types like those inside `defs.json` as input and splits out typescript interfaces for them. Then it is able to decorate the api using those interfaces.

`augment-codec` basically replaces the interfaces automatically generated by `@polkadot/typegen` tool with our custom classes, which may have different or additional methods. I mostly added this with Pioneer in mind, because it would be hard to make it compatible with `augment` (ie. no `isOfType`, `getField` or any other custom methods would be recognized by typescript. There is also no exact `1:1` corelation in getters between `augment` and `augment-codec`, which could cause some hard to spot issues)

**Comment by Lezek123** [comment]
Another thing about `agument` vs `augment-codec` is that:
- If someone uses `defs.json` file to register the types (which should be now be perfectly plausible), `agument` should be 100% compatible with that (this is generally the recommended way of registering types in `@polkadot/api` dependent projects)
- If someone uses `import { types } from '@joystream/types'` to register the types (like pretty much all our projects now), then `augment-codec` should now give 100% compatible typescript augmentation (including custom methods defined in our types classes)

**Comment by mnaamani** [comment]
Excellent. Thanks for the detailed explanation. I feel we should probably compile all this usage and behavior information somewhere. Perhaps in the README of the types package?

The user developer experience is improving with each iteration of this tooling :+1:

**Pull request merged**